### PR TITLE
rfc: introduce concurrent, global scratchpad mode

### PR DESCRIPTION
**Pull request by sbogusev**
This RFC aims to introduce new scratchpad modes to enable more explicit controls for users over (i) memory consumption due to scratchpad, and (ii) thread safety.

[Rendered document](https://github.com/sbogusev/oneDNN/blob/sbogusev/rfcs/rfcs/20200422-concurrent-global-scratchpad-mode/README.md)

### CODE REVIEW

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+user in a particular instance. These usage models are briefly presented in the
+table below:
+
+| Scratchpad mode | Use case | Advantages |
+|-----------------|----------|------------|
+| **library**     | The library allocates scratchpad storage.<br>The mode is chosen by primitive implementation.                    | No API usage<br>     |
+| **concurrent**  | The library allocates scratchpad storage.<br>New scratchpad storage is created for each primitive.              | Thread-safety        |
+| **global**      | The library allocates scratchpad storage.<br>The storage is reused by primitives.                               | Reduced memory usage |
+| **user**        | The user will allocate scratchpad at primitive execution.<br>The scratchpad can be reused at user's discretion. | Full user control    |
+
```
Comment (mgouicem) [inline rfcs/20200422-concurrent-global-scratchpad-mode/README.md:158]: does it makes sense to clean up the naming so that they reflect better the allocation strategy? For example
`private`, `thread_local_pool` and `user` .

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+## Usage models
+
+Since the new scratchpad modes are intended to complement existing one instead
+of replacing them, new usage models are assumed that may be favorable for the
+user in a particular instance. These usage models are briefly presented in the
+table below:
+
+| Scratchpad mode | Use case | Advantages |
+|-----------------|----------|------------|
+| **library**     | The library allocates scratchpad storage.<br>The mode is chosen by primitive implementation.                    | No API usage<br>     |
```
Comment (mgouicem) [inline rfcs/20200422-concurrent-global-scratchpad-mode/README.md:154]: What would be the benefit for a primitive to use private scratchpad (`concurrent`) vs a thread local memory pool 
 (`global`) by default?
And if no benefit, does it make sense to use `library` as a `library_default` mode, and it would be an alias to whatever the compilation flag specified for the default?

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+following is an example of an implementation definition with concurrent
+scratchpad:
+
+~~~cpp
+
+DECLARE_COMMON_PD_T("ncsp_bnorm:any", ncsp_batch_normalization_fwd_t);
+
+~~~
+
+In case a primitive implementation does contain the `USE_GLOBAL_SCRATCHPAD`
```
Comment (mgouicem) [inline rfcs/20200422-concurrent-global-scratchpad-mode/README.md:21]: This seems like a violation of our cmake option. Basically, if the user specifies `DNNL_ENABLE_CONCURRENT_EXEC`, we still create primitives that are not thread safe? Do we know why this appeared?

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 2 above)
Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:154]: The benefit of using `concurrent` is thread safety, enabling the user to create and execute one or more primitives on separate threads. However, if thread safety is not a concern and the user can create and execute primitives on the *same* thread, then the user is encouraged to use `global` scratchpad. The `global` mode is beneficial only in the context of a topology, where multiple primitives may create a scratchpad.

Regarding aliasing, I think `library_default` should remain independent of the compilation, since it emphasizes (at least to me) that the library/developer/primitive implementation decides on the scratchpad mode. Otherwise we may enforce the default to require a compilation flag (probably it's a question of preference).

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 3 above)
Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:21]: From what I understand, `DNNL_ENABLE_CONCURRENT_EXEC` disregards whatever the primitive implementation says in its definition (c.f. [scratchpad](https://github.com/oneapi-src/oneDNN/blob/master/src/common/scratchpad.cpp#L122) )

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 1 above)
Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:158]: I like the idea of replacing these names to something more intuitive. However in this case I think we will have to replace the type name in the library as well. Perhaps we can go with the following:

For the user:
- `concurrent` -> `primitive_private`
- `global` -> `thread_private`

Inside the library:
- `concurrent_scratchpad_t` -> `primitive_private_scratchpad_t`
- `global_scratchpad_t` -> `thread_private_scratchpad_t`

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 3 above)
Comment (mgouicem) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:21]: my bad I had misread the code.

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 2 above)
Comment (mgouicem) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:154]: My question was about the `library` mode. I understand that a user might prefer thread safety (`concurrent`) over memory footprint (`global`).
What I do not understand is when the user does not care (`library` in your proposal), why would the library choose thread safety over memory footprint (and I understand even less that this decision is per primitive).

Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:154]: I think initially it was all about thread safety (primitive private). Then, at some point, global scratchpad was introduced to reduce memory consumption. This led to the existing inconsistency since I don't think there are clear criteria over how to the library decides over scratchpad allocation.

Comment (mgouicem) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:154]: Yes, initially all scratchpads were private, and we added the global scratchpad to save memory footprint (and perf on systems with MCDRAM like Knights Landing).

Back to the initial point, 
1) I do not see when we would like to leave the memory management decision go to the primitives. 
2) we should have a default mode with the library wide policy specified by your proposed build option (needed for the C API).
So I was suggesting to just use your `library` mode as the `default` mode, and its behavior will be the specified by the build option.

Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:154]: Indeed, the primitives will not decide how to manage scratchpad, here I actually meant the current behavior of the library where some primitives specify `GLOBAL` storage and other don't.

For your second point, I agree that the proposed build option should indicate what `scratchpad_mode::library` is. Also, the scope of this RFC covers CPU functionality only, and this mechanism will be documented along with the implementation. There I will also cover the rules for things like `get_scratchpad_mode` and `set_scratchpad_mode`, and how these can be impacted depending mostly on the engine.

Thanks a lot for your comments, I will update the RFC accordingly.

Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md]: I have updated the RFC, please review.

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 1 above)
Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:175]: Updated the RFC, please review.

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+Since the new scratchpad modes are intended to complement existing one instead
+of replacing them, new usage models are assumed to be favorable for the user in
+a particular instance. These usage models are briefly presented in the table
+below:
+
+| Scratchpad mode       | Use case | Advantages |
+|-----------------------|----------|------------|
+| **library**           | The library allocates scratchpad storage.<br>The mode is chosen by primitive implementation.                    | No API usage<br>     |
+| **primitive_private** | The library allocates scratchpad storage.<br>New scratchpad storage is created for each primitive.              | Thread-safety        |
```
Comment (densamoilov) [inline rfcs/20200422-concurrent-global-scratchpad-mode/README.md:140]: I find mentioning thread-safety here confusing because if a primitive was created for this scratchpad mode then it cannot be executed by multiple threads simultaneously. You probably meant that the primitive can be created by one thread, but executed by another. Is that what's implied here?

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+among all or some primitives or not. Therefore, the query cannot guarantee the
+total memory consumption if executing more than one primitive.
+
+
+## Objective
+
+The aim of this proposal is to more provide more explicit controls to the user
+over primitive scratchpad. The main goal is to reduce the total memory usage due
+to scratchpad and/or guarrantee thread safety by using (i) new scratchpad modes
+and (ii) new compilation flag.
```
Comment (densamoilov) [inline rfcs/20200422-concurrent-global-scratchpad-mode/README.md:63]: I'm actually concerned with the flexibility we are going to provide, not the number of options but the limitations that each of them has. It was challenging for users to deal with `user` and `library` modes until @mgouicem documented all the limitations but now we have two more modes. Although the API is common for CPU and GPU the behavior of some policies can potentially differ. For instance, we don't support global scratchpad for GPU so far, but if we decide to implement it there will be no requirement to create and run a primitive in the same thread because there is no sense in making scratchpad tread local for GPU (unless I'm missing something). That means that CPU and GPU primitives that use `thread_private_scratchpad_t` will behave differently and we need to document it.

The ability to create a primitive in one thread but execute it in another was probably useful for frameworks because they could use it to implement their own primitive cache.

Given that the library provides the primitive cache and that it will always be possible to create and run a GPU primitive in different threads, what if we make the `library` mode to always use `thread_private_scratchpad_t` for CPU and `primitive_private_scratchpad_t` for GPU and provide a compile time flag for the case if somebody really needs to use only `primitive_private_scratchpad_t`?
This `library` mode will potentially have the difference in behavior of  CPU and GPU primitive created using the same but we do not introduce new policies.

The general question would be, how important is it to expose all the suggested policies to user? In other words, do we have a clear understanding of the real use cases?

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 4 above)
Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:140]: Indeed that's what I was implying, I will make this more explicit throughout the document.

File: rfcs/20200422-concurrent-global-scratchpad-mode/README.md (hunk 5 above)
Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:63]: Thank you very much for the details.

As it is hopefully described in the background of the RFC, the main motivation came from the fact that only a fraction of primitives on the CPU are currently using a global scratchpad: while user can save memory when using these on a topology level, the property of certain primitives to have a global scratchpad is not transparent for the user (unless you look into the code). This makes the scratchpad policy inconsistent throughout the library, therefore the first idea was to have a toggle for all primitives on CPU. Now the compile flag becomes the only way to manage scratchpad policy, therefore an API would take care of this issue at run-time.

Perhaps, as you suggest, we can simply change the default for CPU primitives to use `thread_private` and GPU to use `primitive_private` (probably also change `thread_private` to `primitive_shared` to account for potentially introducing a global scratchpad on GPU), and tweak it with a compile flag.

The immediate use case would be for frameworks that ensure creating and using CPU primitives on the same thread, therefore `primitive_shared`/`thread_private` would be beneficial for them.

Comment (densamoilov) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:63]: I understand that the main benefit would be a decreased memory consumption. But given that some frameworks already use the `library` mode, that as you mentioned may or may not use the global scratchpad, they are fine with the requirement that the global scratchpad has. So if we switch to the global scratchpad for all CPU primitives then the frameworks will benefit from it without any new policies and changes in their code.
But those frameworks that are not fine with the requirement can just continue using the compile time flag that we already provide to switch to the so called concurrent scratchpad.
To sum it up, with this approach the behavior of the `library` mode stays aligned with the documentation while making it more efficient in terms of memory consumption.

Comment (sbogusev) [reply rfcs/20200422-concurrent-global-scratchpad-mode/README.md:63]: I agree with your idea. In this case I will remove references of new API from this RFC, and will limit the scope to adding support for global scratchpad for all CPU primitives.

**Comment by sbogusev** [comment]
Will close the RFC for now, since it seems I may lack sufficient background for it.
